crates/compiler/load_internal/src/file.rs

```diff
@@ -5505,31 +5163,255 @@ fn canonicalize_and_constrain<'a>(
     }
 }
 
-fn parse<'a>(arena: &'a Bump, header: ModuleHeader<'a>) -> Result<Msg<'a>, LoadingProblem<'a>> {
+fn parse<'a>(
+    arena: &'a Bump,
+    header: ModuleHeader<'a>,
+    module_ids: Arc<Mutex<PackageModuleIds<'a>>>,
+    ident_ids_by_module: SharedIdentIdsByModule,
+) -> Result<Msg<'a>, LoadingProblem<'a>> {
     let mut module_timing = header.module_timing;
     let parse_start = Instant::now();
     let source = header.parse_state.original_bytes();
     let parse_state = header.parse_state;
-    let mut parsed_defs = match module_defs().parse(arena, parse_state.clone(), 0) {
-        Ok((_, success, _state)) => success,
-        Err((_, fail)) => {
+
+    let header_import_defs =
+        roc_parse::ast::Module::header_imports_to_defs(arena, header.header_imports);
+
+    let parsed_defs = match parse_module_defs(arena, parse_state.clone(), header_import_defs) {
+        Ok(success) => success,
+        Err(fail) => {
             return Err(LoadingProblem::ParsingFailed(
                 fail.into_file_error(header.module_path, &parse_state),
             ));
         }
     };
-    for value in header.defined_values.into_iter() {
-        // TODO: should these have a region?
-        parsed_defs.push_value_def(value, Region::zero(), &[], &[]);
-    }
 
     // Record the parse end time once, to avoid checking the time a second time
     // immediately afterward (for the beginning of canonicalization).
     let parse_end = Instant::now();
 
     module_timing.parse_body = parse_end.duration_since(parse_start);
 
-    let imported_modules = header.imported_modules;
+    let mut imported_modules: MutMap<ModuleId, Region> = MutMap::default();
+    let exposed_values = header.header_type.exposed_or_provided_values();
+    let num_exposes = exposed_values.len();
+    let mut deps_by_name: MutMap<PQModuleName, ModuleId> =
+        HashMap::with_capacity_and_hasher(num_exposes, default_hasher());
+
+    match &header.header_type {
+        HeaderType::Platform {
+            opt_app_module_id, ..
+        } => {
+            // Add standard imports, if there is an app module.
+            // (There might not be, e.g. when running `roc check myplatform.roc` or
+            // when generating bindings.)
+            if let Some(app_module_id) = opt_app_module_id {
+                imported_modules.insert(*app_module_id, Region::zero());
+                deps_by_name.insert(
+                    PQModuleName::Unqualified(ModuleName::APP.into()),
+                    *app_module_id,
+                );
+            }
```

Comment:
I'm confused by this. Why is the platform ever implicitly importing the app? Shouldn't that always be explicit?

Reply:
The platform doesn't import the app explicitly.  It's the app that imports the platform and provides values/types to it. This is needed so that those (e.g. `main`) are visible to the platform module while canonicalizing it.

Reply:
Should we rename `imported_modules` to `available_modules` or something like that?

Reply:
Addressed [here](https://github.com/roc-lang/roc/pull/6658/commits/d8d56f6e4927d3695683e3b8b150a000cf082303)